### PR TITLE
tear out bad code, put in good code

### PR DESCRIPTION
**Pull request by willmcgugan**
### What this PR solves
- Replaces ws4py and thread-based re-connect mechanism with Lomond.

### How it is done
- Lots of deleted code, replacing websocket app with Lomond loop.

### What to look out for
- Bugs, and remnants of ws4py code.
- Don't forget to review *wsclient* which has a diff too large to render (its mostly deletions).

### Review
- [X] Ready for review

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -11,8 +11,8 @@ mkdir -p bin
 rm bin/dataplicity
 virtualenv -qq .build
 source .build/bin/activate
-pip -q install pex==1.1.14
+pip -q install pex==1.2.6
 echo building ./bin/dataplicity
-pex dataplicity -r requirements.txt -o bin/dataplicity -m dataplicity.app:main
+pex dataplicity --pre -r requirements.txt -o bin/dataplicity -m dataplicity.app:main
```
Comment (willmcgugan) [inline build.sh]: The `--pre` allows pex to use alpha builds of libs if specified. This was the previous behaviour.

File: dataplicity/m2m/proxy.py [hunk 2]
```diff
@@ -51,6 +51,7 @@ def __init__(self, user=None, group=None, size=None):
             size = [80, 24]
         self.size = size
         self.master_fd = None
+        self.pid = None
```
Comment (willmcgugan) [inline dataplicity/m2m/proxy.py]: Fixes a race condition when a channel is closed before the remote process has spawned. Not sure if this bug had any impact on the end-user. But it did result in a traceback in the logs.

File: dataplicity/m2mmanager.py [hunk 3]
```diff
@@ -71,126 +71,6 @@ def close(self):
         del self.processes[:]
 
 
-class AutoConnectThread(threading.Thread):
```
Comment (willmcgugan) [inline dataplicity/m2mmanager.py]: On reflection, monitoring a thread with another thread is bad idea.

File: dataplicity/jsonrpc.py [hunk 4]
```diff
@@ -204,6 +204,7 @@ def call(self, method, **params):
             "id": call_id
         }
         response_json = self._send(call)
+        log.debug(response_json[:1000])
```
Comment (willmcgugan) [inline dataplicity/jsonrpc.py]: Nice to be sure we are getting the response we are expecting.

File: dataplicity/m2m/remoteprocess.py [hunk 5]
```diff
@@ -71,13 +71,14 @@ def write_master(self, data):
         super(RemoteProcess, self).write_master(data)
 
     def close(self):
-        if not self._closed:
+        if not self._closed and self.pid is not None:
             log.debug('sending kill signal to %r', self)
             # TODO: Implement a non-blocking kill
             os.kill(self.pid, signal.SIGKILL)
+            log.debug('waiting for %r', self)
```
Comment (willmcgugan) [inline dataplicity/m2m/remoteprocess.py]: Just being paranoid, but I'm fairly certain there is nothing to go wrong with killing a child process.

File: dataplicity/m2m/wsclient.py [hunk 6]
```diff
@@ -365,48 +242,36 @@ def hard_close_channels(self):
...
 
     def run(self):
-        sockopt = [(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)]
+        """Main websocket handling loop."""
         try:
-            self.app.run_forever(
-                ssl_options={"cert_reqs": ssl.CERT_NONE}
-            )
+            with self.websocket:
+                for event in persist(self.websocket):
```
Comment (willmcgugan) [inline dataplicity/m2m/wsclient.py]: The call to `persist` replaces what that reconnect thread did.

**Comment by willmcgugan** [comment]
@wildfoundry/web-developers please review

File: dataplicity/m2mmanager.py [hunk 7]
```diff
@@ -200,18 +80,14 @@ def __init__(self, client, url, identity=None):
         self.identity = identity
         self.terminals = {}
         self.notified_identity = None
-        self.connect_thread = AutoConnectThread(self, url, identity=self.identity)
-        self.connect_thread.start()
-
-    @property
-    def m2m_client(self):
-        return self.connect_thread.m2m_client
+        self.m2m_client = WSClient(self, url)
```
Comment (beardygeek) [inline dataplicity/m2mmanager.py]: Does it need to return this?

Comment (beardygeek) [inline dataplicity/m2mmanager.py]: Sorry, ignore me.  I see you deleted the property.  Not reading it right.

**Pull request merged**